Stratis.Bitcoin/Features/Consensus/TimeSyncBehavior.cs

```diff
@@ -0,0 +1,458 @@
+
+        /// <summary>List of timestamp offset samples from peers that our node connected to.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly CircularArray<TimestampOffsetSample> outboundTimestampOffsets;
+
+        /// <summary>List of IP addresses of peers that provided samples in <see cref="inboundSampleSources"/>.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly HashSet<IPAddress> inboundSampleSources;
+
+        /// <summary>List of IP addresses of peers that provided samples in <see cref="outboundTimestampOffsets"/>.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly HashSet<IPAddress> outboundSampleSources;
+
+        /// <summary><c>true</c> if the warning loop has been started, <c>false</c> otherwise.</summary>
+        public bool WarningLoopStarted { get; private set; }
+
+        /// <summary><c>true</c> if the time sync with peers has been switched off, <c>false</c> otherwise.</summary>
+        public bool SwitchedOff { get; private set; }
+
+        /// <summary>
+        /// <c>true</c> if the reason for switching the time sync feature off was that <see cref="timeOffset"/> 
+        /// went over the maximal allowed value, <c>false</c> otherwise.
+        /// </summary>
+        public bool SwitchedOffLimitReached { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="dateTimeProvider">Provider of time functions.</param>
+        /// <param name="nodeLifetime">Global application life cycle control - triggers when application shuts down.</param>
+        /// <param name="asyncLoopFactory">Factory for creating background async loop tasks.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public TimeSyncBehaviorState(IDateTimeProvider dateTimeProvider, INodeLifetime nodeLifetime, IAsyncLoopFactory asyncLoopFactory, ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger(GetType().FullName);
+            this.dateTimeProvider = dateTimeProvider;
+            this.nodeLifetime = nodeLifetime;
+            this.asyncLoopFactory = asyncLoopFactory;
+
+            this.inboundTimestampOffsets = new CircularArray<TimestampOffsetSample>(MaxInboundSamples);
+            this.outboundTimestampOffsets = new CircularArray<TimestampOffsetSample>(MaxOutboundSamples);
+            this.inboundSampleSources = new HashSet<IPAddress>();
+            this.outboundSampleSources = new HashSet<IPAddress>();
+
+            this.timeOffset = TimeSpan.Zero;
+            this.WarningLoopStarted = false;
+            this.SwitchedOff = false;
```

Comment:
What about adding this as a configuration value? -synctime

Reply:
yes, this is exactly prepared for that further development
i did not have enough time here to implement -synctime as a part of this one
but it should now be easy, the logic is basically there